fishPlugins: add sdkman-for-fish

giorgiga opened the pull request:
###### Description of changes

This adds the fish plugin [sdkman-for-fish](https://github.com/reitzig/sdkman-for-fish), which enables [SDKMAN!](https://sdkman.io/) for fish users.

Since SDKMAN! normally only works on bash and similar shells, sdkman-for-fish is quite important for java developers who use the fish shell (an alternative plugin is https://github.com/deather/omf-sdk).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

giorgiga commented:
I wrote this mimicking other fish plugins, but **I did not test it**.

I couldn't figure out how plugins should be tested... I run `nix-build -A fish` (but I don't think that actually tested anything) and the `nixpkgs-review ` command above.

I put my info as a maintainer (should I add myself to [the maintainer list](https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix) instead?). I don't particularly care about being a maintainer (and, given the nature of sdkman-for-fish there is very little to maintain anyways) so feel free to take this over if you wish; if you don't, I have no issues updating this from time to time, but I might end up crying for help if actual nix code needs to be written in the future.

Please instruct me on how to proceed.

giorgiga commented:
*Forgot to mention:*

SDKMAN! is not available on nix, but:
- it is supposed to be installed in $HOME
- it isn't a requirement for this plugin (the plugin will download it if needed)

natsukium commented:
Thanks for your contribution!

First, please add yourself to the maintainer list.
see https://nixos.org/manual/nixpkgs/stable/#var-meta-maintainers and git log.

Second, this package is "fishPlugin", so it should be added as `fishPlugins.sdkman`.
You also need to register it in `pkgs/shells/fish/plugins/default.nix`.
see, https://nixos.org/manual/nixpkgs/stable/#sec-fish-plugins-pkg

Now you can build this package with `nix-build -A fishPlugins.sdkman`.

Also, the binary this plugin downloads probably won't work on NixOS, so something needs to be done.

giorgiga committed:
maintainers: add giorgiga

giorgiga committed:
fishPlugins.sdkman-for-fish: init at 1.4.0

Co-authored-by: OTABI Tomoya <tomoya.otabi@gmail.com>
Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

giorgiga commented:
Thanks @natsukium !

- [x] added myself as a maintaner
- [x] renamed plugin to `sdkman-for-fish`
- [x] fixed plugin and registered it in fishPlugin
- [x] run `nix-build lib/tests/maintainers.nix` and `nix-build -A fishPlugins.sdkman-for-fish`

As for running SDKMAN! in NixOS, that most probably won't be a problem as SDKMAN! is all shell scripts (no binaries).

natsukium commented on pkgs/shells/fish/plugins/sdkman-for-fish.nix:
```diff
@@ -0,0 +1,20 @@
+{ lib, buildFishPlugin, fetchFromGitHub }:
+
+buildFishPlugin rec {
+  pname = "sdkman";
```
```suggestion
  pname = "sdkman-for-fish";
```
The `pname` should be the same as the upstream.
https://nixos.org/manual/nixpkgs/unstable/#sec-package-naming

giorgiga replied:
nice catch - I missed that

natsukium commented on pkgs/shells/fish/plugins/sdkman-for-fish.nix:
```diff
@@ -0,0 +1,20 @@
+    owner = "reitzig";
+    repo = "sdkman-for-fish";
+    rev = "v${version}";
+    sha256 = "sha256-bfWQ2al0Xy8bnJt5euziHz/+qhyri4qWy47VDoPwQcg=";
```
```suggestion
    hash = "sha256-bfWQ2al0Xy8bnJt5euziHz/+qhyri4qWy47VDoPwQcg=";
```
SRI hash is preferable.

natsukium reviewed (commented):
The upstream has a test set, so we need to run it with `cucumber` in `checkPhase`.
However, some features depend on `SDKMAN!` and it seems necessary to disable them because of network access.

giorgiga commented:
About the tests from upstream, is it by policy that they must be run?

I mean, it's not like those are integration tests that should be run to ensure that the project builds/works when its dependencies are provided by nix (there are no dependencies)

natsukium commented:
I would like to run the tests except for those that cannot be run due to networking, DB access, etc.
However, I also think that the tests are trivial since this package does not depend on other packages and only provides commands, as you say.

natsukium commented:
Would you please squash the commits to "maintainers: add giorgiga" and "fishPlugins: add sdkman-for-fish"?

giorgiga commented:
Commits squashed.

I also re-run `nix-build lib/tests/maintainers.nix` and `nix-build -A fishPlugins.sdkman-for-fish`, just in case.

The resulting files seem fine:
````
> tree /nix/store/w7mxx3rw2wc9qzwlpphla7sxgf769z0a-fishplugin-sdkman-for-fish-1.4.0
/nix/store/w7mxx3rw2wc9qzwlpphla7sxgf769z0a-fishplugin-sdkman-for-fish-1.4.0
└── share
    └── fish
        ├── vendor_completions.d
        │   └── sdk.fish
        ├── vendor_conf.d
        │   └── sdk.fish
        └── vendor_functions.d
            └── sdk.fish
````

I don't think it's strictly needed given installing this is just a matter of copying files to the right place, but... do you happen to know if I can somehow reference this PR from my `home.nix` to test it? (sorry to be a newb)

natsukium commented:
Thanks!

I think it is easy to use wrapFish.
https://github.com/NixOS/nixpkgs/blob/master/doc/builders/packages/fish.section.md

Or if you manage home.nix with flakes, you can try nixpkgs url to this local directory, or `-I nixpkgs=${path}` for nix-shell.

giorgiga commented:
`nix-shell -p fish fishPlugins.sdkman-for-fish -I nixpkgs=. --run fish`  just uses the `sdkman-for-fish` from my `~/.config/fish` ... but trying a different plugin doesn't seem to get me anywhere either (I tried `z` which I have not installed and the `z` command was not available inside nix-shell).

Testing before it's available in unstable seems to be way above my current nix skills.

natsukium commented:
Would you please try `nix-shell -p 'wrapFish { pluginPkgs = with fishPlugins; [ sdkman-for-fish ]; } -I nixpkgs=. --run fish`?

giorgiga commented:
Thank you: I would never have managed to write that :sweat_smile:  (well, I will be able to in due time)

Anyway, it works!

I uninstalled both the plugin and sdkman from my home, reinstalled sdkman via that shell and (after restarting the shell, but that's expected) `sdk list java` would list the available java packages as expected.

natsukium commented:
Sorry, could you change the second commit message to `fishPlugins.sdkman-for-fish: init at 1.4.0`?
My previous request was wrong. I apologize.

giorgiga commented:
No problem and no need to apologize! I am thankful for your guidance.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/2157

SuperSandro2000 commented on pkgs/shells/fish/plugins/sdkman-for-fish.nix:
```diff
@@ -0,0 +1,20 @@
+  };
+
+  meta = with lib; {
+    description = " Adds support for SDKMAN! to fish";
```
```suggestion
    description = "Adds support for SDKMAN! to fish";
```

giorgiga replied:
Nice catch!

Fixed with https://github.com/NixOS/nixpkgs/pull/227125/commits/3fee00dd9c9716a9e7c2ca1823ec688e0ccec5e5
